src/python/pants/engine/internals/scheduler.py

```diff
@@ -693,29 +712,65 @@ def register_task(rule: TaskRule) -> None:
         for awaitable in rule.awaitables:
             unions = [t for t in awaitable.input_types if is_union(t)]
             if len(unions) == 1:
-                # Register the union by recording a copy of the Get for each union member.
                 union = unions[0]
-                in_scope_types = union_in_scope_types(union)
-                assert in_scope_types is not None
-                for union_member in union_membership.get(union):
-                    native_engine.tasks_add_get_union(
-                        tasks,
-                        awaitable.output_type,
-                        tuple(union_member if t == union else t for t in awaitable.input_types),
-                        in_scope_types,
-                    )
+                if awaitable.rule_id:
+                    if rule_id_to_rule[awaitable.rule_id].polymorphic:
+                        # This is a polymorphic call-by-name. Compute its vtable data, i.e., a list
+                        # of pairs (union member type, id of implementation rule for that type).
+                        member_rule_type_pairs = base_type_to_member_rule_type_pairs.get(union)
+                        vtable_entries: list[tuple[type[Any], str]] = []
+                        for member_rule, member_type in member_rule_type_pairs or []:
+                            # If a rule has a union member as a param, and returns the relevant
+                            # output type, then we take it to be the implementation of the
+                            # polymorphic rule for the union member.
+                            if member_rule.output_type == awaitable.output_type:
```

Comment:
Are collisions silent?

Reply:
They are silent, and I suppose they are possible (would probably lead to a solver error today, but would be possible once we're fully CBN). Should probably check for that...